ctk/common/src/main/kotlin/org/codice/compliance/verification/core/SubjectComparisonVerifier.kt

```diff
@@ -0,0 +1,267 @@
+                            "equal to the content of the other [${id2.textContent}]",
+                    node = response)
+        }
+    }
+
+    /**
+     * This function verifies that the Subjects from the Response **strongly match** the Subject
+     * from the AuthnRequest.
+     *
+     * NOTE: This method does not test to make sure the Response Subject's identifier's Format is
+     * equal to an optional NameIDPolicyFormat from the AuthnRequest. That is tested in
+     * NameIDPolicyVerifier
+     *
+     * NOTE: This method does not actually resolve who the principal of the subject is. It only
+     * compares the given Subject values to each other. In order to fully test this function we
+     * need to resolve the Subjects to a principal.
+     */
+    fun verifySubjectsMatchAuthnRequest() {
+
+        val requestSubject = request?.children("Subject")?.firstOrNull() ?: return
+        val requestId = requestSubject.getId()
+        val requestConfirmations = requestSubject.getSubjectConfirmations()
+
+        if (requestId == null && requestConfirmations == null) return
+
+        val nameIdPolicyFormat =
+                request.children("NameIDPolicy").firstOrNull()?.filteredFormatValue()
+
+        response.recursiveChildren("Assertion")
+                .flatMap { it.children(SUBJECT) }
+                .forEach { resSubject ->
+
+                    // Verify ids match
+                    requestId?.let { reqId ->
+                        resSubject.getId()?.let { resId ->
+                            verifyIdsMatchAuthnRequest(reqId, resId, nameIdPolicyFormat)
+                        } ?: throw SAMLComplianceException.create(SAMLCore_3_3_4_b1,
+                                SAMLCore_3_4_1_4_a1,
+                                message = "One of the Response's Subjects contained no identifier",
+                                node = resSubject)
+                    }
+
+                    // Verify SubjectConfirmations match
+                    requestConfirmations?.let { reqConfirmations ->
+                        resSubject.getSubjectConfirmations()?.let { resConfirmations ->
+                            verifyConfirmationsMatchAuthnRequest(reqConfirmations,
+                                    resConfirmations)
+                        } ?: throw SAMLComplianceException.create(SAMLCore_3_3_4_c1,
+                                SAMLCore_3_4_1_4_a1,
+                                message = "One of the Response's Subjects contained no " +
+                                        "SubjectConfirmations.",
+                                node = resSubject)
+                    }
+                }
+    }
+
+    private fun verifyIdsMatchAuthnRequest(
+            requestId: Node,
+            responseId: Node,
+            nameIdPolicyFormat: String?) {
+
+        verifyIdAttributesMatchAuthnRequest(requestId, responseId, nameIdPolicyFormat)
+        verifyIdContentsMatchAuthnRequest(requestId, responseId)
+    }
```

Comment:
✏️ This private function feels unnecessary to me. In its one usage above, I think it's just as clear to inline the call to the `verifyIdAttributesMatchAuthnRequest` and `verifyIdContentsMatchAuthnRequest` methods.